src/Messages/OWSMessageSender.m

```diff
@@ -228,6 +236,9 @@ - (void)tryWithRemainingRetries:(NSUInteger)remainingRetries
             [self tryWithRemainingRetries:remainingRetries - 1];
         } else {
             DDLogWarn(@"%@ Too many failures. Giving up sending.", self.tag);
+
+            [self.messageSender saveMessage:self.message withError:error];
```

Comment:
This also needs to happen for terminal failures (like SN changes). 

You could copy paste it above, or probably better, move to the failureHandler wrapper in the `initWithMessage `.